<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>General · CHOPOFF.jl</title><meta name="title" content="General · CHOPOFF.jl"/><meta property="og:title" content="General · CHOPOFF.jl"/><meta property="twitter:title" content="General · CHOPOFF.jl"/><meta name="description" content="Documentation for CHOPOFF.jl."/><meta property="og:description" content="Documentation for CHOPOFF.jl."/><meta property="twitter:description" content="Documentation for CHOPOFF.jl."/><script data-outdated-warner src="assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="assets/documenter.js"></script><script src="search_index.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="assets/themeswap.js"></script><link href="assets/theme.css" rel="stylesheet" type="text/css"/></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href><img class="docs-light-only" src="assets/logo.png" alt="CHOPOFF.jl logo"/><img class="docs-dark-only" src="assets/logo-dark.png" alt="CHOPOFF.jl logo"/></a><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li class="is-active"><a class="tocitem" href>General</a><ul class="internal"><li><a class="tocitem" href="#About"><span>About</span></a></li><li><a class="tocitem" href="#Requirements"><span>Requirements</span></a></li><li><a class="tocitem" href="#Standalone-application"><span>Standalone application</span></a></li><li><a class="tocitem" href="#No-build-application"><span>No-build application</span></a></li><li><a class="tocitem" href="#Quick-Use"><span>Quick Use</span></a></li><li><a class="tocitem" href="#Support"><span>Support</span></a></li><li><a class="tocitem" href="#LICENSE"><span>LICENSE</span></a></li></ul></li><li><span class="tocitem">API</span><ul><li><a class="tocitem" href="abstract_gRNA/">Abstract gRNA</a></li><li><a class="tocitem" href="find_potential_ot/">Find potential off-targets</a></li><li><a class="tocitem" href="align_gRNA/">Align gRNA and off-target</a></li><li><a class="tocitem" href="alignment_free/">Alignment-free filters for gRNAs</a></li><li><a class="tocitem" href="find_ot/">Find all off-targets</a></li><li><a class="tocitem" href="utils/">Utils</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>General</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>General</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/JokingHero/CHOPOFF.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="CHOPOFF.jl"><a class="docs-heading-anchor" href="#CHOPOFF.jl">CHOPOFF.jl</a><a id="CHOPOFF.jl-1"></a><a class="docs-heading-anchor-permalink" href="#CHOPOFF.jl" title="Permalink"></a></h1><h2 id="About"><a class="docs-heading-anchor" href="#About">About</a><a id="About-1"></a><a class="docs-heading-anchor-permalink" href="#About" title="Permalink"></a></h2><p>Uncompromising finding of CRISPR off-targets:</p><ul><li>many fast alignment algorithms optimized specifically for CRISPR</li><li>search for larger distances allowing for mismatches and bulges</li><li>support for ambiguous bases</li><li>arbitrarily large genomes</li><li>VCF support - with multiple overlapping SNPs</li><li>near-instant alignment-free off-target filtering</li><li>pruning of off-targets by their location (remove overlapping, competing off-targets)</li><li>extensively tested</li><li>full framework that can be extended for your own algorithms with ease</li></ul><h2 id="Requirements"><a class="docs-heading-anchor" href="#Requirements">Requirements</a><a id="Requirements-1"></a><a class="docs-heading-anchor-permalink" href="#Requirements" title="Permalink"></a></h2><ul><li><p>Some algorithms generate as many files as there are prefixes (e.g. for prefix 7 - this will make 4^7 - 16384 files). This strategy allows us to operate the searches independently on multiple cores and not get throttled when querying large number of the guides. However, some systems have artificial limits on the number of open files, for example in Ubuntu &#39;ulimit -n&#39; will show the limit. Increase the limits, if it creates problems for you.</p></li><li><p>When using many cores for building the indexes - you have to have around ~1 GB of RAM per thread.</p></li></ul><h2 id="Standalone-application"><a class="docs-heading-anchor" href="#Standalone-application">Standalone application</a><a id="Standalone-application-1"></a><a class="docs-heading-anchor-permalink" href="#Standalone-application" title="Permalink"></a></h2><p>It is possible to build CHOPOFF into standalone application - which includes all dependencies and Julia into one compiled software. This is <strong>recommended</strong> method for using of CHOPOFF when you are not a developer. If you know how to code in Julia, you might make use of the whole framework using CHOPOFF as a package.</p><p>To build a standalone application run <code>./build_standalone.sh</code> script from the main directory. Script will produce binary in a &quot;build&quot; folder. Then you can run from inside that folder <code>./bin/CHOPOFF --help</code>. To learn about building a database run <code>./bin/CHOPOFF build --help</code> and to use existing database check out <code>./bin/CHOPOFF search --help</code>. It is possible to skip testing + precompile step to speed up the build process with <code>./build_standalone.sh --noprecompile</code>.</p><p>You can alternatively download the latest release from the releases&#39; page on the GitHub.</p><p>When using application as self-contained compiled software, you can control number of cores by setting <code>JULIA_NUM_THREADS</code> environment variable.</p><p><strong>Example commands for using standalone</strong></p><p>Building of <code>prefixHashDB</code> database for standard Cas9 <code>--motif</code> with support for up to levenshtein distance 3 <code>--distance</code> for an example genome using 10 threads.</p><pre><code class="language-bash hljs">export JULIA_NUM_THREADS=10  
EXAMPLE_GENOME=&quot;./test/sample_data/genome/semirandom.fa&quot;
CHOPOFF build --name Cas9_hg38 --genome &quot;$EXAMPLE_GENOME&quot; -o out_dir/phDB_16_3/ --distance 3 --motif Cas9 prefixHashDB</code></pre><p>Searching of above database for all off-targets for guides listed in <code>--guides</code> up to the 2 levenshtein distance <code>--distance</code> using 15 threads, writing the results into <code>--output</code> file. Because <code>--early_stopping</code> argument is not supplied below, by default <code>prefixHashDB</code> will search for up to 1e6 off-targets per guide per distance. Pay attention that default guides for the Cas9, are 20bp long, as can be inspected in the example file.</p><pre><code class="language-bash hljs">export JULIA_NUM_THREADS=15  
EXAMPLE_GUIDES=&quot;./test/sample_data/guides.txt&quot;
CHOPOFF search --database phDB_16_3/ --guides &quot;$EXAMPLE_GUIDES&quot; --output out_dir/phDB_16_2.csv --distance 2 prefixHashDB</code></pre><h2 id="No-build-application"><a class="docs-heading-anchor" href="#No-build-application">No-build application</a><a id="No-build-application-1"></a><a class="docs-heading-anchor-permalink" href="#No-build-application" title="Permalink"></a></h2><p>Run CHOPOFF package as an <strong>application</strong>, without building first. From the directory of the package run:</p><pre><code class="language-bash hljs">julia --threads 4 --project=&quot;.&quot; ./src/CHOPOFF.jl --help</code></pre><h2 id="Quick-Use"><a class="docs-heading-anchor" href="#Quick-Use">Quick Use</a><a id="Quick-Use-1"></a><a class="docs-heading-anchor-permalink" href="#Quick-Use" title="Permalink"></a></h2><p>For search of off-targets you have a couple of options:</p><ul><li><strong><a href="find_ot/#CHOPOFF.search_prefixHashDB">prefixHashDB</a></strong> - the fastest, we apply hashes to symbolic alignments for fast filtering of OTs  </li><li><a href="find_ot/#CHOPOFF.search_linearDB">linearDB</a> - most rigorously tested</li><li><a href="find_ot/#CHOPOFF.search_motifDB">motifDB</a> - on top of linearDB we apply pigeonhole principle like filter which you can adjust</li><li><a href="find_ot/#CHOPOFF.search_treeDB">treeDB</a> - will work best for longer gRNAs, uses vantage point for filteirng</li><li><a href="find_ot/#CHOPOFF.search_fmiDB">fmiDB</a> - the smallest file size, very fast, but only for distances ≤ 2</li><li><a href="find_ot/#CHOPOFF.search_binaryFuseFilterDB">binaryFuseFilterDB</a> - uses hashing on top of FM-index</li></ul><p>If you would like to <strong>filter</strong> or <strong>rank</strong> gRNAs to only those that are most likely off-target free you want to use <a href="alignment_free/#CHOPOFF.search_hashDB">hashDB</a> or slower and much larger, but less probabilistic <a href="alignment_free/#CHOPOFF.search_dictDB">dictDB</a>.</p><p>For VCF file support use <a href="find_ot/#CHOPOFF.search_vcfDB">vcfDB</a>.  For use of the framework as a Julia package consult the documentation - Public Interface section.</p><h2 id="Support"><a class="docs-heading-anchor" href="#Support">Support</a><a id="Support-1"></a><a class="docs-heading-anchor-permalink" href="#Support" title="Permalink"></a></h2><p>You can buy me a <a href="https://www.buymeacoffee.com/kornellabun">coffee</a> to show some love and appreciation!</p><img src="./assets/bmc_qr.png" width="25%"/><h2 id="LICENSE"><a class="docs-heading-anchor" href="#LICENSE">LICENSE</a><a id="LICENSE-1"></a><a class="docs-heading-anchor-permalink" href="#LICENSE" title="Permalink"></a></h2><p>Copyright © 2022 Kornel Labun</p><p>License for non-commercial applications is aGPL-3.0.  For commercial applications you should acquire permission or licensing contract.</p><p><a href="https://tldrlegal.com/license/gnu-affero-general-public-license-v3-(agpl-3.0)">https://tldrlegal.com/license/gnu-affero-general-public-license-v3-(agpl-3.0)</a></p><p>This program is free software for non-commercial applications:  you can redistribute it and/or modify it under the terms of the GNU Affero General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.</p><p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more details.</p><p>You should have received a copy of the GNU Affero General Public License along with this program inside LICENSE file.  If not, see <a href="https://www.gnu.org/licenses/">https://www.gnu.org/licenses/</a>.</p></article><nav class="docs-footer"><a class="docs-footer-nextpage" href="abstract_gRNA/">Abstract gRNA »</a><div class="flexbox-break"></div><p class="footer-message">Made with ♥ by Kornel Labun.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.3.0 on <span class="colophon-date" title="Monday 12 May 2025 09:40">Monday 12 May 2025</span>. Using Julia version 1.8.5.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
