<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Find potential off-targets · CHOPOFF.jl</title><meta name="title" content="Find potential off-targets · CHOPOFF.jl"/><meta property="og:title" content="Find potential off-targets · CHOPOFF.jl"/><meta property="twitter:title" content="Find potential off-targets · CHOPOFF.jl"/><meta name="description" content="Documentation for CHOPOFF.jl."/><meta property="og:description" content="Documentation for CHOPOFF.jl."/><meta property="twitter:description" content="Documentation for CHOPOFF.jl."/><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../search_index.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script><link href="../assets/theme.css" rel="stylesheet" type="text/css"/></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../"><img class="docs-light-only" src="../assets/logo.png" alt="CHOPOFF.jl logo"/><img class="docs-dark-only" src="../assets/logo-dark.png" alt="CHOPOFF.jl logo"/></a><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../">General</a></li><li><span class="tocitem">API</span><ul><li><a class="tocitem" href="../abstract_gRNA/">Abstract gRNA</a></li><li class="is-active"><a class="tocitem" href>Find potential off-targets</a></li><li><a class="tocitem" href="../align_gRNA/">Align gRNA and off-target</a></li><li><a class="tocitem" href="../alignment_free/">Alignment-free filters for gRNAs</a></li><li><a class="tocitem" href="../find_ot/">Find all off-targets</a></li><li><a class="tocitem" href="../utils/">Utils</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">API</a></li><li class="is-active"><a href>Find potential off-targets</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Find potential off-targets</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/JokingHero/CHOPOFF.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="Find-potential-off-targets"><a class="docs-heading-anchor" href="#Find-potential-off-targets">Find potential off-targets</a><a id="Find-potential-off-targets-1"></a><a class="docs-heading-anchor-permalink" href="#Find-potential-off-targets" title="Permalink"></a></h1><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="CHOPOFF.DBInfo" href="#CHOPOFF.DBInfo"><code>CHOPOFF.DBInfo</code></a> — <span class="docstring-category">Type</span></header><section><div><p><code>DBInfo(filepath::String, name::String, motif::Motif; vcf_filepath::String = &quot;&quot;)</code></p><p>Motif defines what genome file is being used for the searches.</p><p><strong>Arguments</strong></p><p><code>filepath</code> - Path to the genome file, if file is fasta (ends with .fa or .fasta or .fna)  make sure you also have fasta index file with extension .fai. Alternatively, you can use .2bit  genome file.  </p><p><code>name</code> - Your name for this instance of DBInfo: the genome with connection to the motif and vcf file.</p><p><code>motif</code>   - <code>Motif</code> object defining search parameters</p><p><code>vcf_filepath</code>  - Optional. Path to the VCF file to include in the searches.</p><p>Alignments will be performed from opposite to the extension direction (which is defined by extend5).</p><p><strong>Examples</strong></p><pre><code class="language-julia hljs"># use CHOPOFF example genome
genome = joinpath(vcat(splitpath(dirname(pathof(CHOPOFF)))[1:end-1], 
    &quot;test&quot;, &quot;sample_data&quot;, &quot;genome&quot;, &quot;semirandom.fa&quot;))
# construct example DBInfo
DBInfo(genome, &quot;Cas9_semirandom_noVCF&quot;, Motif(&quot;Cas9&quot;))</code></pre></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="CHOPOFF.gatherofftargets!" href="#CHOPOFF.gatherofftargets!"><code>CHOPOFF.gatherofftargets!</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">function gatherofftargets!(
    output::T,
    dbi::DBInfo;
    remove_pam::Bool = true,
    normalize::Bool = true,
    restrict_to_len::Union{Nothing, Int64} = nothing) where {T&lt;:Union{Vector{String}, Vector{UInt64}, Vector{UInt128}}}</code></pre><p>Gathers all off-targets that conform to the given <code>dbi</code> Motif.</p><p>This function appends to the <code>output</code> during the run, however it will also return all ambiguous  guides in return object. We can use UInt64 and UInt128 to compress space that the gRNAs use. When using  large genomes or non-specific PAMs you might run out of memory when using this function.</p><p>remove<em>pam - whether PAM sequence should be removed normalize - whether all guides should be flipped into PAMseqEXT e.g. GGn-20N-3bp restrict</em>to<em>len - will restrict the guides to be of specific lengths, smaller than the initial motif     this includes/excludes PAM based on remove</em>pam as remove_pam is applied before the length restriction</p><p><strong>Examples</strong></p><pre><code class="language-julia hljs"># use CHOPOFF example genome
genome = joinpath(
    vcat(
        splitpath(dirname(pathof(CHOPOFF)))[1:end-1], 
        &quot;test&quot;, &quot;sample_data&quot;, &quot;genome&quot;, &quot;semirandom.fa&quot;))
# construct example DBInfo
dbi = DBInfo(genome, &quot;Cas9_semirandom_noVCF&quot;, Motif(&quot;Cas9&quot;))
# finally gather all off-targets
guides = Vector{String}()
ambig = gatherofftargets!(guides, dbi)

# here in the format of UInt64 encoding
guides2 = Vector{UInt64}()
ambig2 = gatherofftargets!(guides2, dbi)
guide_with_extension_len = length_noPAM(dbi.motif) + dbi.motif.distance

# transform UInt64 to LongDNA and String
guides2 = String.(LongDNA{4}.(guides2, guide_with_extension_len))</code></pre></div></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../abstract_gRNA/">« Abstract gRNA</a><a class="docs-footer-nextpage" href="../align_gRNA/">Align gRNA and off-target »</a><div class="flexbox-break"></div><p class="footer-message">Made with ♥ by Kornel Labun.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.3.0 on <span class="colophon-date" title="Monday 12 May 2025 13:07">Monday 12 May 2025</span>. Using Julia version 1.8.5.</p></section><footer class="modal-card-foot"></footer></div></div></div></body><div data-docstringscollapsed="true"></div></html>
